File: src/core/adb-monitor.js
const _ = require('lodash');
const Adb = require('./adb');
const DeviceService = require('../service/device.service');

const WATCH_RETRY = 10000;
const DEVICE_PORT = 5555;
let localPortCounter = 7777;

class AdbMonitor {

    constructor(deviceService, adb) {
        if (!deviceService || !(deviceService instanceof DeviceService))
            throw new Error('AdbMonitor must have a valid deviceService');
        if (!adb || !(adb instanceof Adb))
            throw new Error('AdbMonitor must have a valid adb');
        this.deviceService = deviceService;
        this.adb = adb
    }

    start() {
        this.adb.killServer()
            .then(() => this.adb.startServer())
            .then(() => this.scanDevices());
    }

    scanDevices() {
        console.log('=========================================================================');
        console.log('============================== scanDevices ==============================');
        console.log('=========================================================================');
        this.adb.devices()
            .then(devices => {
                console.log('AdbMonitor => all connected devices:');
                this.logDevices(devices);

                const preparedDevices = this.deviceService.devices();

                const newDevices = _.differenceBy(devices, preparedDevices, device => device.id);
                console.log("AdbMonitor => connected now:");
                this.logDevices(newDevices);

                const removedDevices = _.differenceBy(preparedDevices, devices, device => device.id);
                console.log("AdbMonitor => removed now:");
                this.logDevices(removedDevices);

                const newPreparedDevices = _.differenceBy(preparedDevices, removedDevices, device => device.id);
                console.log("AdbMonitor => preparedDevices: ");
                this.logDevices(newPreparedDevices);

                this.deviceService.refreshDevices(newPreparedDevices);

                return Promise.all(newDevices.map(device =>
                    //only add to prepared if setup is ok!
                    this.setupDevice(device, 250)
                        .then(() => this.deviceService.create(device))))
            })
            .catch(error => console.error('AdbMonitor => error:', error))
            .then(() => {
                console.log('AdbMonitor => scheduling another scan in ' + WATCH_RETRY + 'ms');
                setTimeout(() => this.scanDevices(), WATCH_RETRY)
            });
    }

    /** @returns {Promise} */
    setupDevice(device, timeout) {
        console.log('AdbMonitor => setupDevice:', device.id);
        let localPort = localPortCounter++;
        return this.adb.tcpIpFor(device.id, timeout)
            .then(() => this.adb.forwardFor(device.id, localPort, DEVICE_PORT))
            // .then(() => this.adb.forwardList(device.id))
            .then(() => this.adb.connect(device.id, localPort))
            .then(() => this.adb.deviceModelFor(device.id)
                .then(model => this.adb.deviceManufacturerFor(device.id)
                    .then(manufacturer => {
                        device.setManufacturer(manufacturer);
                        device.setModel(model);
                        device.setPort(localPort);
                    })));
    }

    logDevices(devices) {
        if (!_.isEmpty(devices)) {
            console.log("[");
            devices.forEach(d => console.log('  ' + JSON.stringify(d)));
            console.log(']');
        } else {
            console.log("[]");
        }
    }

}

module.exports = AdbMonitor;

File: src/main/kotlin/org/andcoe/adf/core/AdbMonitor.kt
package org.andcoe.adf.core

import org.andcoe.adf.devices.DeviceId
import org.andcoe.adf.devices.DevicesService

class AdbMonitor(
    private val devicesService: DevicesService,
    private val adb: Adb
) {

    var localPortCounter = 7777

    private companion object {
        const val DEVICE_PORT = 5555
        const val SCAN_RETRY_EVERY = 2_000L
    }

    fun startScanning() {
        restartAdb()
        while (true) {
            println("=========================================================================")
            println("============================== scanDevices ==============================")
            println("=========================================================================")
            val connectedDevices = adb.devices()
            refreshDevicesWith(connectedDevices)
            println("AdbMonitor => scheduling another scan in ${SCAN_RETRY_EVERY}ms")
            Thread.sleep(SCAN_RETRY_EVERY)
        }
    }

    fun restartAdb() {
        adb.killServer()
        adb.startServer()
    }

    fun refreshDevicesWith(connectedDevices: List<DeviceId>) {
        println("AdbMonitor => connected devices: $connectedDevices")

        val preparedDevices = devicesService.devices()

        val removedDevices: List<DeviceId> = preparedDevices.keys.filter { !connectedDevices.contains(it) }
        println("AdbMonitor => removed now: $removedDevices")

        preparedDevices
            .filter { removedDevices.contains(it.key) }
            .map { devicesService.delete(it.key) }

        val newDevices: List<DeviceId> = connectedDevices.filter { !preparedDevices.containsKey(it) }
        println("AdbMonitor => connected now: $newDevices")

        newDevices.forEach {
            val adbDevice = setupDevice(it.id)
            devicesService.create(adbDevice)
        }

        println("AdbMonitor => prepared devices: ${devicesService.devices().map { it.key }}")
    }

    private fun setupDevice(deviceId: String): AdbDevice {
        val localPort = localPortCounter++
        adb.tcpIpFor(deviceId)
        adb.forwardFor(deviceId, localPort, DEVICE_PORT)
        adb.connect(deviceId, localPort)

        val model = adb.modelFor(deviceId)
        val manufacturer = adb.manufacturerFor(deviceId)
        val androidVersion = adb.androidVersionFor(deviceId)
        val apiLevel = adb.apiLevelFor(deviceId)
        return AdbDevice(deviceId, model, manufacturer, androidVersion, apiLevel, localPort.toString())
    }

}